vm/src/builtins/enumerate.rs

```diff
@@ -76,32 +82,90 @@ impl PyValue for PyReverseSequenceIterator {
 
 #[pyimpl(with(PyIter))]
 impl PyReverseSequenceIterator {
-    pub fn new(obj: PyObjectRef, len: isize) -> Self {
+    pub fn new(obj: PyObjectRef, len: usize) -> Self {
         Self {
-            position: AtomicCell::new(len - 1),
+            position: AtomicCell::new(len.saturating_sub(1)),
+            status: AtomicCell::new(if len == 0 { Exhausted } else { Active }),
             obj,
         }
     }
 
     #[pymethod(magic)]
-    fn length_hint(&self) -> PyResult<isize> {
-        Ok(self.position.load() + 1)
+    fn length_hint(&self, vm: &VirtualMachine) -> PyResult<usize> {
+        Ok(match self.status.load() {
+            Active => {
+                let position = self.position.load();
+                if position > vm.obj_len(&self.obj)? {
+                    0
+                } else {
+                    position + 1
+                }
+            }
+            Exhausted => 0,
+        })
+    }
+
+    #[pymethod(magic)]
+    fn setstate(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        // When we're exhausted, just return.
+        if let Exhausted = self.status.load() {
+            return Ok(());
+        }
+        let len = vm.obj_len(&self.obj)?;
+        let pos = state
+            .payload::<PyInt>()
+            .ok_or_else(|| vm.new_type_error("an integer is required.".to_owned()))?;
+        let pos = std::cmp::min(
+            try_to_primitive(pos.as_bigint(), vm).unwrap_or(0),
+            len.saturating_sub(1),
+        );
+        self.position.store(pos);
+        Ok(())
+    }
+
+    #[pymethod(magic)]
+    fn reduce(&self, vm: &VirtualMachine) -> PyResult {
+        let iter = vm.get_attribute(vm.builtins.clone(), "reversed")?;
+        Ok(vm.ctx.new_tuple(match self.status.load() {
+            Exhausted => {
+                let obj_type = self.obj.clone_class();
+                let empty_obj = if let Some(ref tp_new) = obj_type.slots.new {
+                    tp_new(
+                        vm,
+                        FuncArgs {
+                            args: vec![obj_type.clone().into_object()],
+                            kwargs: IndexMap::new(),
+                        },
+                    )?
+                } else {
+                    vm.ctx.new_list(vec![])
+                };
```

Comment:
An empty tuple is returned when the iterator is exhausted. See i.e:

```python
r = reversed("foo"); list(r); r.__reduce__()       # (reversed, ((),))
r = reversed((1, 2, 3)); list(r); r.__reduce__()   # (reversed, ((),))

class Foo:
    def __len__(self):
        return 10
    def __getitem__(self, i):
        if i < 10:
            return i
        raise StopIteration

r = reversed(Foo()); list(r); r.__reduce__()  # (reversed, ((),))
```

Reply:
Oh. you are right. I thought it must return empty object with given type as deque object in cpython work like below.
```python
>>> from collections import deque
>>> a = reversed(deque())
>>> list(a)
[]
>>> a.__reduce__()
(<class '_collections._deque_reverse_iterator'>, (deque([]), 0))
```
But there exist specialized methods for deque, not a general one. This is my miss. I sent v2 code.
Thanks!